### PR TITLE
Handle references to missing contracts

### LINKED ISSUE #8773: Handle references to non-existent contract ids in Daml Script dumps
There are two cases where we might see a reference to a contract id that does not exist:

1. The contract was divulged. There isn’t anything we can do about this for now but ledger API changes are planned to allow us to fetch that contract. For now, we should just emit a reasonably useful error suggesting the cause. Another option could be to take the contract id as a parameter to the daml script dump and leave it to users to reconstruct it.
2. The contract id is just made up. As long as you don’t fetch the contract id, you can use an arbitrary string (with some formatting restrictions) in contract ids. This isn’t a useful case but we probably still want to handle it in some form at the very least by producing a reasonable error or maybe again by taking it as parameter.

Note that we cannot distinguish 1 and 2.

### PR DESCRIPTION
**Pull request by aherrmann-da**
Closes #8773 

- The generated Daml script now takes an argument of type `Args` which is a record with the fields `parties` and `contracts`. `contracts` is a mapping from old contract ids of missing contracts to replacement contract ids. `parties` is the record of parties from before.
- Daml ledger export no longer fails if it encounters unknown contract ids. Instead, it emits a lookup of this contract id in the `contracts` argument in the generated Daml script code.
- Daml ledger export now generates a default `args.json` file that configures the default parties and contracts mapping. The default parties are the same parties used to generate the ledger export. The default contracts mapping maps the unknown contract ids to themselves. Users can modify `args.json` or provide a separate input file to configure the generated Daml script.
- The generated Daml script no longer defines `allocateParties` or `testExport` as these are effectively replaced by `args.json`.
- The generated `daml.yaml` file sets `parties`, `init-script` and `script-options` such that `daml start` will run the export script on the default `args.json` file.
- The Daml ledger export documentation is updated.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by cocreature** [comment]
> The generated Daml script no longer defines allocateParties or testExport as these are effectively replaced by args.json.

`testExport` is great to test it easily in Daml Studio or even against a real ledger to make sure it runs correctly and maybe start playing around with it. It would be great if we could still have something that allows for this.

**Review by cocreature — APPROVED** [review]
Nice work! I’d like to preserve some form of the `allocateParties` stuff for IDE integration. Perfectly fine, if it falls apart if you have missing contract ids.

**Comment by aherrmann-da** [comment]
> > The generated Daml script no longer defines allocateParties or testExport as these are effectively replaced by args.json.
> 
> `testExport` is great to test it easily in Daml Studio or even against a real ledger to make sure it runs correctly and maybe start playing around with it. It would be great if we could still have something that allows for this.

Testing against a real ledger should be easy with the help of the `args.json` file. The generated `daml.yaml` is configured such that `daml start` works out of the box.

> I’d like to preserve some form of the `allocateParties` stuff for IDE integration. Perfectly fine, if it falls apart if you have missing contract ids.

Fair enough. Yes, in it's current form `testExport` only works if there are no missing contract-ids. Two ways we could address this (in a follow-up):
- Add a Daml script primitive to conjure up contract ids, e.g. `unsafeCreateContractId : Text -> ContractId a`.
- Add a mechanism to read `args.json` from Daml script, e.g. `readJSON : FilePath -> Script (Option a)` or `scriptWithArgs : FilePath -> (a -> Script b) -> Script b`

I'll revive `testExport` for now.

**Comment by cocreature** [comment]
> Testing against a real ledger should be easy with the help of the `args.json` file. The generated `daml.yaml` is configured such that `daml start` works out of the box.

It only works if you work against a ledger that has implicit party allocation (and potentially it also requires an empty ledger). Otherwise, I need to allocate the parties separately which can be a bit annoying.

File: daml-script/export/src/main/scala/com/daml/script/export/Encode.scala [hunk 1]
```diff
@@ -25,35 +38,30 @@ private[export] object Encode {
       Doc.hardLine +
       encodePartyType(export.partyMap) /
       Doc.hardLine +
-      encodeAllocateParties(export.partyMap) /
+      Doc.text("type Contracts = DA.TextMap.TextMap (ContractId ())") /
```
[the remark this answers is not in the document]
Comment (aherrmann-da) [reply daml-script/export/src/main/scala/com/daml/script/export/Encode.scala:54]: Done. I've added a snippet how one could provide a replacement in Daml script if one had a contract deployed already.

File: daml-script/export/src/main/scala/com/daml/script/export/Encode.scala [hunk 2]
```diff
@@ -25,35 +38,30 @@ private[export] object Encode {
...
-        Doc.text("parties <- allocateParties") /
-        Doc.text("export parties")).hang(2) /
+      Doc.text("getContract : DA.Stack.HasCallStack => Text -> Contracts -> ContractId a") /
+      (Doc.text("getContract old contracts =") /
+        (Doc.text("case DA.TextMap.lookup old contracts of") /
+          Doc.text("None -> error (\"Missing contract id \" <> old)") /
+          Doc.text("Some new -> coerceContractId new")).nested(2)).nested(2) /
+      Doc.hardLine +
+      (Doc.text("data Args = Args with") /
+        Doc.text("parties : Parties") /
```
[the remark this answers is not in the document]
Comment (aherrmann-da) [reply daml-script/export/src/main/scala/com/daml/script/export/Encode.scala:65]: done

**Pull request merged**